### ISSUE TITLE
Inconsistent display of group membership badges

### ISSUE THREAD

**Issue by val-kulkov** [NodeBB/NodeBB#10921]
### NodeBB version

v2.5.3

### NodeBB git hash

cf6e8101e86e0d5f91b491421e6477a2b74d3a81

### NodeJS version

v18.7.0

### Installed NodeBB plugins

Active plugins:
	* nodebb-plugin-2factor@5.0.2 (installed, disabled)
	* nodebb-plugin-category-notifications@4.0.2 (installed, enabled)
	* nodebb-plugin-composer-default@9.1.1 (installed, enabled)
	* nodebb-plugin-dbsearch@5.1.5 (installed, enabled)
	* nodebb-plugin-embed@3.0.19 (installed, enabled)
	* nodebb-plugin-emoji@4.0.4 (installed, enabled)
	* nodebb-plugin-emoji-android@3.0.0 (installed, enabled)
	* nodebb-plugin-lightgallery@1.0.18 (installed, enabled)
	* nodebb-plugin-markdown@10.1.0 (installed, enabled)
	* nodebb-plugin-mentions@3.0.11 (installed, enabled)
	* nodebb-plugin-poll@2.0.6 (installed, enabled)
	* nodebb-plugin-spam-be-gone@1.0.0 (installed, disabled)
	* nodebb-rewards-essentials@0.2.1 (installed, enabled)
	* nodebb-theme-lavender@6.0.0 (installed, disabled)
	* nodebb-theme-persona@12.1.1 (installed, enabled)
	* nodebb-theme-slick@2.0.2 (installed, disabled)
	* nodebb-theme-vanilla@12.1.18 (installed, disabled)
	* nodebb-widget-essentials@6.0.0 (installed, enabled)

### Database type

PostgreSQL

### Database version

v14.3

### Exact steps to cause this issue

We have two standard system groups, "administrators" and "Global Moderators", and one custom group. Let's call the custom group "CG". Almost all users are members of CG. "Show Badge" is enabled for all three groups.

Issue 1: the group membership badge appears correctly for most of the members of CG, but not all. Unexplicably, the CG membership badge does not appear for at least one user. That user is only a member of CG. The group membership setup of that user is no different from the group membership setup of the majority of other users.

Issue 2: no group membership badge appears for users who are members of the "administrators" group. No badge at all despite the concurrent membership in "administrators", "Global Moderators" and "GC", with "Show Badge" enabled for all groups.

I suspect that something might be wrong in the [getGroupsMap function](https://github.com/NodeBB/NodeBB/blob/aad0a618c86f45055827dada020bad11b041bd0d/src/posts/user.js#L83-L99), however I have no way of checking this at present because I am unfamiliar with the NodeBB internal architecture and the code. I considered debugging getGroupsMap to see why I observe no badge for at least one user, and why members of the "administrators" group do not get any badges at all, but I found that in order to understand the documentation for debugging, I need to at least have a basic understanding of the NodeBB internal structure. For a solo project, that's too much for me. However, if someone could provide me enough info to setup a debugging environment, I would be happy to invest time into debugging this issue.

### What you expected

Consistent, predictable behaviour

### What happened instead

See above

**Comment by julianlam** [comment]
Users are able to select which badges they elect to show from their user settings. That is likely the culprit here.

If a user has elected to show no badges, then them being added to another group doesn't make the badge show up, as the settings have been explicitly saved as "show no badge".

Does that help?

**Comment by val-kulkov** [comment]
Thank you, this may explain the absence of a badge for that particular user, but it does not explain why no badges at all show up for members of the "administrators" group. My account belongs to the "administrators", "Global Moderators" and "CG" groups, and I cannot get any badge displayed next to my name.

Additionally, I have just checked my user settings. I could not find any option to control whether a badge is shown or not shown. If there is one, could you point me in the right direction? Thank you.

**Comment by julianlam** [comment]
The option is located in User > Edit, titled "Group Title"

![image](https://user-images.githubusercontent.com/923011/192030370-478641aa-4581-4f6e-8661-8f15eb2be251.png)

I will admit it is not the most intuitive.

**Comment by val-kulkov** [comment]
Thank you! I see this option now and I agree, the location is not intuitive.

I am going to play with this setting and see what difference it makes (and what records in the database are changed). If this gives me enough information to fix the issue in our particular case, that's great! But then I think I might redefine this issue as "in case of multiple group memberships, the default is setting for 'Group Title' should be something other than 'No group title'." I will think about it.

**Comment by julianlam** [comment]
> the default is setting for 'Group Title' should be something other than 'No group title'."

It is, as far as I know. If the user has never edited their account, then `groupTitle` is undefined, and the default behaviour is to select the first group title they are a member of.

**However**, if they've edited their account at least once, then the setting is persisted going forward, and the pre-selected option is the first one: "No Group Title".

**Comment by val-kulkov** [comment]
@julianlam : yes, I agree, this may indeed be the crux of the "inconsistency" - I am currently looking into it, and I will report my findings once I have completed my investigation.

I can see however that the current situation is not ideal in the sense that any user who has edited their account at least once may be exposed to the confusion described above and to this obscure configuration setting. I am beginning to think that any change in group membership for a user should trigger an automatic addition/deletion of that group from the user's groupTitle value, if one is present.

**Comment by val-kulkov** [comment]
@julianlam : it is indeed the "No group title" selection that causes the observed issue. Technically, it is not a bug. Rather, it is a non-intuitive workflow that produces unexpected results. This, in turn, confuses both users and admins. The existence of this issue is an evidence of such confusion.

I propose that on the user's Edit Profile page, there is a "Hide all group memberships" checkbox, off by default, and the "No group title" item is eliminated from the list of groups to display. Then, the value of groupTitle must change on a new group addition in all cases, not only when the current value of groupTitle is undefined. Here is why.

THE CURRENT WORKFLOW

The default value of groupTitle is undefined. As soon as a user edits his/her profile for the first time, the value of groupTitle becomes defined. If, at the time of the first profile editing, the user is not a member of any groups, the value of groupTitle becomes "" (empty string). This prevents display of any group membership in the future when the user acquires such group membership. What makes this hiding of group membership non-intuitive and confusing is the fact that the user did not turn his/her mind to this hiding of membership at the time of the first profile editing. This is so not only because "No group title" made sense to the user at the time when he/she was not a member of any group, but also because "No group title" is the only option in the list and the user cannot opt out of it even if the user is smart enough to figure out that "No group title" really means "hide all group memberships".

Once the value of groupTitle is defined, an addition of group membership for a user does not change the value of groupTitle. I argue that it is not intuitive. If, for example, a user is a member of multiple groups and the value of his/her groupTitle reflects that (showing multiple group memberships), a membership in a new group should should change the value of groupTitle correspondingly. At present, the user has to manually edit his/her profile to add the new group to the list of groups in groupTitle.

Notably, a removal of a user from a group does change the value of groupTitle. The removal of the last group membership sets the value of groupTitle to "" (empty string), which in effects locks in the "hide all group memberships" mode for the user.

THE PROPOSED WORKFLOW

Every group membership addition/deletion should change the value of groupTitle accordingly. The group memberships should be displayed according to the value of groupTitle unless the value of the "hide all group memberships" checkbox is defined and has the value "on".

The default value of the "hide all group memberships" checkbox should be undefined or "off". This way, only those users who turn their mind to setting the value of this checkbox to "on" will have their group memberships hidden. In my view, this would be the intuitive and expected result from an explicit action by a user, removing the current confusion.

Those users who do not turn on the "hide all group memberships" option would still be able to manually select which group memberships they want to have displayed.

The migration from the current workflow to the proposed one would involve setting the value of the "hide all group memberships" checkbox to "on" for those users who currently have groupTitle == "", and setting the value of groupTitle to their current group memberships.

I think I want to rename the title of this issue from the current "Inconsistent display of group membership badges" to "non-intuitive workflow caused by setting value of groupTitle to empty string". I am not sure that this issue should be classifed as "bug", but I am not sure that it is a "feature request" either. In any case, I probably cannot reclassify this issue by myself.

**Comment by val-kulkov** [comment]
I also suggest that in Settings -> Groups, the default value of "Allow Multiple Badges" is set to "on". When it is "off", as it is now, the default behaviour is to display only one group membership even if a user is a member of multiple groups. In my view, this is not intuitive. If a member of a group is also an admin with elevated privileges, I would expect to see that the user is both a member and an admin.

**Comment by mySebbe** [comment]
The two main UX complaints are no longer current: profile editing now exposes per-badge show/hide plus ordering, and `allowMultipleBadges` defaults to enabled. The remaining behavior is narrower: once `groupTitle` has any persisted value (including empty), joining a new group does not automatically add that badge. I suggest closing this original inconsistency issue and opening a focused product issue only if new memberships should opt in automatically despite a user's saved badge selection.
